model/src/main/kotlin/OrtResult.kt

```diff
@@ -62,7 +62,14 @@ data class OrtResult(
      * An [EvaluatorRun] containing details about the evaluation that was run using the result from [scanner] as
      * input. Can be null if no evaluation was run.
      */
-    val evaluator: EvaluatorRun? = null
+    val evaluator: EvaluatorRun? = null,
+
+    /**
+     * User defined labels associated to this result. Labels are not used by ORT itself, but can be used in parts of ORT
+     * which are customizable by the user, for example in evaluator rules or in the notice reporter.
+     */
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    val labels: Map<String, String> = emptyMap()
```

Comment:
We already have `val data: CustomData = mutableMapOf()` (which deliberately is not part of the primary constructor, so it's not taken into account in the data classes' `hashCode()` and `copy()`). Wouldn't it make more sense to just use that?

Reply:
What was actually the reason for d824a74? I don't remember why we didn't want `data` to be respected by the autogenerated functions.

The reason I didn't use it is that these labels will become part of the evaluator and reporter API, and I don't want these to depend on a property that's intended for third-party tools.

Reply:
> What was actually the reason for [d824a74](https://github.com/oss-review-toolkit/ort/commit/d824a743b501301caf2c6264938ed990841dd9a4)? I don't remember why we didn't want `data` to be respected by the autogenerated functions.

Because in several code locations we rely on OrtResults to be equal despite their custom data, so we must exclude custom data from `hashCode()` / `equals()`.

> The reason I didn't use it is that these labels will become part of the evaluator and reporter API, and I don't want these to depend on a property that's intended for third-party tools.

Maybe we should relax that initial idea a bit by saying that the custom data field is for, well, custom data that *might* come from / is handled by third-party tools, but also ORT itself may make optional use of it.

Originally I though you only want to introduce command line options to set custom data elegant via the CLI instead of e.g. patching ORT result files.

Anyway, the "custom data" and "labels" approach are awfully similar, and I'd very much like to consolidate the two if possible.

Reply:
As discussed offline the labels will replace the data property after the migration is done. I have added that to the commit message of the first commit. Regarding `hashCode()`/`equals()` we agreed to keep `labels` in the constructor and try if it causes any issues.